### PR TITLE
Error message details dialog

### PR DESCRIPTION
**Pull request by jhass**
Includes and supersedes #22

Sorry for the drive by commits :)

### CODE REVIEW

File: lib/src/app_auth.dart [hunk 1]
```diff
@@ -108,14 +108,20 @@ class AppAuth {
...
-      if (e.code.startsWith("timeout")) {
+      if (e.code.startsWith("timeout_")) {
         throw TimeoutException(e.message);
       }
 
+      if (e.message?.toLowerCase()?.contains("network") == true) { // probably bad network
+        throw AuthorizationFailedException("bad_network", e.message, e.details);
+      }
+
+      if (e.code.startsWith("failed_")) {
```
Comment (tclaus) [inline lib/src/app_auth.dart:128]: Shouldn't this be moved above the 'network' heuristics? I still think a` ..message.contains.('network')`.. is much weaker than testing for a known Code `beginsWith('failure..')`

File: lib/src/app_auth.dart [hunk 2]
```diff
@@ -108,14 +108,20 @@ class AppAuth {
...
+        // Catches all failed_discovery (Unknown service/API), failed_register, failed_auth. etc
+        throw AuthorizationFailedException(e.code, e.message, e.details);
+      }
+
```
Comment (tclaus) [inline lib/src/app_auth.dart:132]: If executions falls trough the exception tests - is it a good idea to log this somewhere?

File: lib/src/localizations.dart [hunk 3]
```diff
@@ -150,9 +178,32 @@ class InsporationLocalizations {
...
-    'Timeout while trying to authenticate, are you sure your pod supports the API?',
+    'Timeout while trying to authorize, are you sure your pod supports the API?', // TODO update after AppAuth-Android#611
     name: 'errorSignInTimeout',
-    desc: 'Error message after authenticating timed out',
+    desc: 'Error message after authorization timed out',
+    locale: localeName
+  );
+
+  String errorAuthorizationFailed(String userId) => Intl.message(
+      'Could not authorize to $userId, is it spelled correctly, is your network working and is your pod running the latest development snapshot?', // TODO update message after diaspora 0.8 release
```
Comment (tclaus) [inline lib/src/localizations.dart:188]: We talked before about pasting private data in auch a error message. Should the userID still be part of the error message? User might not want so send this to a developer with their personal username inside it?.

File: lib/src/app_auth.dart [hunk 4]
```diff
@@ -108,14 +108,20 @@ class AppAuth {
 
       return tokens["accessToken"];
     } on PlatformException catch (e) {
-      if (e.message?.toLowerCase()?.contains("network") == true) {
-        throw e.message; // probably bad network
-      }
 
-      if (e.code.startsWith("timeout")) {
+      if (e.code.startsWith("timeout_")) {
```
Comment (tclaus) [inline lib/src/app_auth.dart:120]: In iOS side the Timeout code is not specified: just "timeout" - is it needed to specify the timeouts by scope, if we now handle the message and details?

Comment (tclaus) [reply lib/src/app_auth.dart:120]: Chnanged swift part to 'timeout_' - code

**Review by tclaus — CHANGES_REQUESTED** [review]
Some minor changes. 
It feels uncommon to see a warning above the textfield and the login-button.
The visual flow seems to be disturbed.  Can the message be between the button and a possible list of other usernames? 
Bildschirmfoto 2020-11-14 um 11 26 17

File: lib/src/app_auth.dart (hunk 1 above)
Comment (jhass) [reply lib/src/app_auth.dart:128]: All the network failures come back with a `failed_` code, at least for Android. The code only tells a story about *where* something failed, not *how*. The "network" heuristic tries to reason about the *how*, so it has to come before we swallow all the *where* cases.

File: lib/src/localizations.dart (hunk 3 above)
Comment (jhass) [reply lib/src/localizations.dart:188]: This part is not copied to the clipboard by the app. 

Generally this is quite useful information for debugging issues as you mentioned yourself before. Also I think it's quite obvious that it's there when sharing this with us, and easy enough to edit out.

File: lib/src/app_auth.dart (hunk 2 above)
Comment (jhass) [reply lib/src/app_auth.dart:132]: I felt this easy enough to see all information from within the app. In this case and just this case the user facing message will be `errorUnexpectedErrorOnAuthorization` and the detailed exception is available behind the show trace functionality.

File: lib/src/app_auth.dart (hunk 4 above)
Comment (jhass) [reply lib/src/app_auth.dart:120]: Might be a nice idea to do anyhow, might prove to be easier to understand where an error came from like that :)

**Comment by jhass** [comment]
> Some minor changes.
> It feels uncommon to see a warning above the textfield and the login-button.
> The visual flow seems to be disturbed. Can the message be between the button and a possible list of other usernames?
> Bildschirmfoto 2020-11-14 um 11 26 17

I played with various positions and this one felt the least worst. Placing it between the button and the list or between the logo and the form field felt even more disturbing. You can easily play with this by moving the `ErrorMessage` widget to different positions within the `Column` widget.

By the way looking at this, I guess I forgot to populate the error details parameter on the RPC bus with a formatted full trace for the iOS side of things. Do you want to take a look at that?

**Comment by tclaus** [comment]
Yep. I'll take a look.

**Pull request merged**